zebra/zebra_rib.c

```diff
@@ -1386,8 +1389,11 @@ static void rib_process_del_fib(struct zebra_vrf *zvrf, struct route_node *rn,
 	}
 
 	/* Update nexthop for route, reset changed flag. */
-	nexthop_active_update(rn, old, 1);
-	UNSET_FLAG(old->status, ROUTE_ENTRY_CHANGED);
+	if (!nexthop_active_update(rn, old, 1) &&
```

Comment:
my understanding of the problem is that the kernel route has a nexthop who's interface is down now and we are not getting a callback about it.  What happens when that interface comes back up, is the kernel route restored?

Reply:
when interface comes back up, the kernel route did not restore the route, because the route was flushed. 
in the kernel I tested ( ubuntu 16.04), the route is flushed when nexthop is unreachable.

Reply:
I'm really confused about what we are trying to fix then.  If the route is flushed are we not getting a RTM_DELROUTE for it?

Reply:
yes. we are not getting RTM_DELROUTE
we received the interface down event.
but we did not receive the RTM_DELROUTE event

Reply:
as remind, we are trying to fix #3152

Reply:
Confirmed that when an interface goes down, there is no DELROUTE for IPv4 routes, they're silently deleted from the kernel.

But I'm not sure the fix is correct?

Reply:
I'm not really sure where the fix should go either, have been tossing it around in my mind for a while now and not really coming up with something better.  Let me spend a few minutes here thinking about it.

Reply:
so I like the fix because it's targeted to just kernel routes, in my mind it's not a lot different than the `here be dragons` bit of code.  I think I'm ok with this.

Reply:
I'm going to add a comment above this and merge